Reactive streams integration conformance to TCK verified & fixed

elizarov opened the pull request:
* Throw NPE on null subscriber
* Throw IAE when requested n <= 0 elements
* Do not sent any notification after cancel()

elizarov commented on reactive/kotlinx-coroutines-reactive/test/ReactiveStreamTckTest.kt:
```diff
@@ -0,0 +1,209 @@
+import org.reactivestreams.tck.*
+
+@RunWith(Parameterized::class)
+class ReactiveStreamTckTest(
```
I've tried to, but could not figure out in a reasonable time how to make it run in Gradle. Moreover, parameterized tests in TestNG require XML files (or so it seems), which I would rather not be introducing.

elizarov commented on reactive/kotlinx-coroutines-reactive/src/Publish.kt:
```diff
@@ -42,6 +42,8 @@ public fun <T> CoroutineScope.publish(
     @BuilderInference block: suspend ProducerScope<T>.() -> Unit
 ): Publisher<T> = Publisher { subscriber ->
+    // specification requires NPE on null subscriber
+    if (subscriber == null) throw NullPointerException("subscriber")
```
Fixed.

JakeWharton commented on reactive/kotlinx-coroutines-reactive/test/ReactiveStreamTckTest.kt:
```diff
@@ -0,0 +1,209 @@
+    companion object {
+        @Parameterized.Parameters(name = "{0}")
+        @JvmStatic
+        fun params(): Collection<Array<Any>> = Dispatcher.values().map { arrayOf<Any>(it) }
```
You don't need the nested array here. See "Tests with single parameter" on https://junit.org/junit4/javadoc/4.12/org/junit/runners/Parameterized.html

elizarov replied:
Good to know. Thanks!

elizarov commented on reactive/kotlinx-coroutines-reactive/src/Publish.kt:
```diff
@@ -148,6 +153,12 @@ private class PublisherCoroutine<in T>(
+                // Specification requires that after cancellation requested we don't call onXXX
+                if (cancelled) {
+                    // but we cannot just ignore exception so we handle it
+                    if (cause != null) handleCoroutineException(context, cause, this)
```
Indeed. The whole code is wrong with respect to modern state of error-handling. It signals completion "onCancellation" which is simply a wrong way to do, since the final exception is not there yet. 

I've created a test that reproduces this problem (`PublishTest.testHandleFailureAfterCancel`) and fixed it by handling errors after completion.

elizarov commented:
There are still at least two problems with this code:

1.  It is racy. Exceptions should not handled in `onCompletedExceptionally` (which I do now) since by that time the coroutine & its parent can already be _complete_ and this miss exception (exception should be handled in `handleJobException` which is fired before final state).

2. This code checks for `isActive` and if it is false (the coroutine was cancelled) it takes its (non-final) exception and uses it to report onComplete/onError. It is wrong, because the code might still fail while terminating.

qwwdfsad commented:
To avoid races you can propagate `onError` (and/or check cancelled flag) from `handleCoroutineException`. But anyway, please provide a stress test for a race on `isActive`.

Also, please rename `onCancellation` to `onCancelling`, the current name is confusing and does not match our terminology.

elizarov commented:
Fixed everything. Added tests, fixed various races. Now it waits for all child exceptions to aggregate, cancel parents on exception and never looses exceptions even without parent.
